### PR TITLE
PCHR-3928: Notifications for Onboarding Wizard

### PR DESCRIPTION
**Pull request by powdevel**
## Overview
Sends notifications about user data changes when a user submits the Onboarding Wizard. Changes on data are detected comparing the view results before submission is saved and after submission is inserted. If changes are detected an email is sent to the admin if the admin has previously chosen to be notified (setting inside Onboarding Wizard configuration form).

## New Classes and Updated Classes

2 Classes were created:
- `Drupal\civihr_employee_portal\View\ViewsResultDiff` obtains differences between view results.
- `Drupal\civihr_employee_portal\Mail\ContactDetailsChangesMail` sends notifications to the admin with the changes of user data.

`Drupal\civihr_employee_portal\Mail\AbstractDrupalSystemMail` class was refactored and classes that extend this class, `AccountActivatedSystemMail` and `ContactDetailsChangesMail` were updated accordingly.

## Additions
- **7 Views** were created on `./views/views_export/notifications` to obtain data about changes due to Onboarding submissions. These views also define the HTML output on notification emails, more precisely labels and values of the changed fields and the title of the section as the view title.

- **6 templates** inside `./templates/smarty/notifications` were created to wrap the results of the differences and build an entire notification email.

- **3 functions** were created on `Drupal\civihr_employee_portal\Helpers\HelperClass`, 2 of them are related to avoid using IDs for values like _Location Types_ or _Phone Types_, because these IDs could vary between installations.

## Implemented Hooks
- `civihr_employee_portal_mail` _hook_mail_
- `civihr_employee_portal_preprocess_views_view_fields` _hook_preprocess_views_view_fields_
- `civihr_employee_portal_webform_submission_presave` _hook_submission_presave_

## Before

No notifications were sent on data changes related to contact due to Onboarding submissions.

## After
The following are examples notifications taken from Chrome PC in a Gmail account:

image

image

### CODE REVIEW

**Comment by mickadoo** [comment]
@powdevel This PR is targeting `PCHR-3747-Notifications-for-Onboarding-Wizard` but the [PCHR-3928 ticket](https://compucorp.atlassian.net/browse/PCHR-3928) is a technical task of parent [PCHR-3748](https://compucorp.atlassian.net/browse/PCHR-3748). Can you explain the structure here?

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+   *     the string which could have underscores or not
+   *
+   * @return string
+   *     the string without the number at the end,
+   *     if no number found at the end returns the same string
+   */
+  private static function removeLastNumberOfString($string) {
+    $pieces = explode('_', $string);
+    $lastIndex = count($pieces) - 1;
+    return $lastIndex > 0 && intval(array_pop($pieces)) ? implode('_', $pieces) : $string;
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: space before `return`, also this long line of logic would be a lot more readable with separate variables like

```
$pieces = explode('_', $string);
$lastIndex = count($pieces) - 1;
$hasEnoughParts === ($lastIndex > 0);
if (!$hasEnoughParts) {
  return $string;
}
$lastPart = array_pop($pieces);
$lastPartIsInt === is_int($lastPart);

return  $lastPartIsInt ? implode('_', $pieces) : $string;
```

Although I think if I was writing this I'd probably avoid expensive `explode` and `implode` calls and just use some function that used `strrpos` and `substr`

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+   * where "base name of the table" is the name of the table without the number
+   * at the end
+   *
+   * @param string $tableName
+   *     this is the table name exported by views
+   *
+   * @return string
+   *     the name of the existing table if found or an empty string
+   */
+  private static function getCiviCRMtableName($tableName) {
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: I don't know what this function does, it takes and argument `$tableName` and returns `CiviCRMtableName`. Have you checked if there's an easier way to get the table names than using the global `$databases` variable, have you looked at `CRM_Core_DAO_AllCoreTables`?

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    global $databases;
+    static $tableNamesMap = [];
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: I don't think we should use static function variables, this seems like a good situation to use just a regular class property and fetch it just once. Does this class need to be static or can we make it non-static?

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    if (!$tableNamesMap) {
+      $tablesWithIds = array_keys($databases['default']['default']['prefix']);
+      foreach ($tablesWithIds as $tableWithId) {
+        $tableNamesMap[self::removeLastNumberOfString($tableWithId)] = $tableWithId;
+      }
+    }
+
+    $exportedTableBaseName = self::removeLastNumberOfString($tableName);
+    $tableFound = isset($tableNamesMap[$exportedTableBaseName]);
+    return $tableFound ? $tableNamesMap[$exportedTableBaseName] : '';
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: space before return

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  /**
+   * To obtain the prefix for a table in the current installation
+   *
+   * @param string $tableName
+   *     the name of the table
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: 2 spaces before description

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+   *
+   * @return string
+   *     the prefix for the table
+   */
+  private static function getPrefix($tableName) {
+    global $databases;
+    return $databases['default']['default']['prefix'][$tableName];
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: Please avoid using global variables if you can, it's better to use a class that allows access to this in case something changes with the underlying implementation later.

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  /**
+   * Gets the columns from a table keyed by the name of the column without
+   * the sequential at the end
+   *
+   * @param string $table
+   *     the name of the table
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: spaces before description

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+   *
+   * @return array
+   *     the names of the columns of the table
+   */
+  private static function getColumnsFromTable($table) {
+    static $columns = [];
+    if (!isset( $columns[$table])) {
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: extra space before `$columns`

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+   *
+   * In the array where "group" key is defined:
+   * - The value will be an array of "views" settings, as defined above
+   *
+   * @var array
+   */
+  public $viewsSettings = [];
+
+  /**
+   * usually will hold data after a form submission but previous to save or insert
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: This has the same description as `$oldData` - is this correct?

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+   * @param  array $oldData
+   *     previous result from view
+   * @param  array $newData
+   *     a recent result from view
+   * @param  array $changes
+   *     the difference between $oldData and $newData
+   *
+   * @return stdClass $this
+   */
+  private function compareDataRecursive(&$oldData, &$newData, &$changes) {
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: Why are $oldData and $newData passed by reference if their values are just checked in this function?

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    return $this;
+  }
+
+  /**
+   * To retrieve the current data of the views configured to diff
+   * this function will be called by storeNewData or storeOldData
+   * to be able to store the data produced by the views for later
+   * comparison
+   *
+   * @return array
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: missing `$param`

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+    $data = [];
+    $settings = !$settings ? $this->viewsSettings : $settings;
+
+    foreach ($settings as $setting) {
+      // if it is a group then apply recursion
+      if (isset($setting['group'])) {
+        $data[] = ['group' => $this->getCurrentData($setting['group'])];
+      } else if (isset($setting['view'])) {
+        // if it is a view it must be previewed to get the results
+        $view = views_get_view($setting['view']);
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: Is there a way we can create separate classes for comparing and fetching the data? Ideally the diff generator would just receive 2 sets of data and generate the diff without any drupal function calls.

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+        // Setting Exposed Filters
+        // backing up $_GET
+        $_GETbkp = $_GET;
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: This seems really messy, is there no other way to get view data without doing this?

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+        // add exposed filter keys to $_GET
+        $_GET = isset($setting['exposedFilters']) ? $setting['exposedFilters'] + $_GET : $_GET ;
+        $view->preview($setting['displayId'], $setting['args']);
+        // reverting $_GET to backed state once the view is executed
+        $_GET = $_GETbkp;
+
+        // views which are empty will not have fieldsMarkup
+        $view->fieldsMarkup = isset($view->fieldsMarkup) ? $view->fieldsMarkup : [];
+        $currentViewData = [];
+        $fieldKeyForRows = isset($setting['fieldAsKeyForRows']) ? $setting['fieldAsKeyForRows'] : false;
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: boolean case

File: civihr_employee_portal/src/View/ViewsResultDiff.php [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+    return $data;
+  }
+
+  /**
+   * Stores data on property newData
+   *
+   * @return ViewsResultDiff
+   */
+  public function storeNewData() {
+    $this->newData = $this->getCurrentData();
```
Comment (mickadoo) [inline civihr_employee_portal/src/View/ViewsResultDiff.php]: the whole distinction between `storeNewData` and `storeOldData` is confusing, this whole class is doing a lot of stuff, I'd suggest breaking it down into separate services like
```
ViewsDiffGenerator::getDiff($firstDataSet, $secondDataSet);
ViewsDataFetcher::getCurrentData($all, $params, $that, $are, $required);
```

Also the whole view setting structure is so complicated that it doesn't make any sense when I'm looking at it in the constructor. Maybe you could consider creating an object for this, in general this class is just hard to follow - it would be better to break it into more specific classes. I think diff generation is an ideal candidate for unit testing but I don't see any tests in this PR at all. With smaller, more specific classes we can write clear tests for them that don't require Drupal function calls.

File: civihr_employee_portal/templates/smarty/notifications/html.tpl [hunk 16]
```diff
@@ -0,0 +1,151 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
```
Comment (mickadoo) [inline civihr_employee_portal/templates/smarty/notifications/html.tpl]: Where did you get the code for this?

File: civihr_employee_portal/views/views_export/notifications/mydetails_notifications__emergency_contacts.inc [hunk 17]
```diff
@@ -0,0 +1,202 @@
+<?php
+
+$view = new view();
```
Comment (mickadoo) [inline civihr_employee_portal/views/views_export/notifications/mydetails_notifications__emergency_contacts.inc]: I wasn't clear just from reading the PR description, why are we creating all these new views? Is it just to generate the email content? If that's the case then why not just use smarty templates if we already have all the data required to generate them?

File: civihr_employee_portal/civihr_employee_portal.module [hunk 18]
```diff
@@ -5136,13 +5181,16 @@ function _civihr_employee_portal_should_do_onboarding($account) {
  * @param \stdClass $submission
  */
 function civihr_employee_portal_webform_submission_insert($node, $submission) {
-  $nodeTitle = property_exists($node, 'title') ? $node->title : NULL;
-  $onboardingTitle = OnboardingWebForm::NAME;
-  $isOnboardingForm = $nodeTitle === $onboardingTitle;
-
-  if ($isOnboardingForm) {
+  if ($node->title === OnboardingWebForm::NAME) {
```
[the remark this answers is not in the document]
Comment (powdevel) [reply civihr_employee_portal/civihr_employee_portal.module]: @mickadoo in my experience, and please tell me if I'm wrong, I haven't seen a node without a title property. I think this check it not needed.

File: civihr_employee_portal/src/View/ViewsResultDiff.php (hunk 13 above)
Comment (powdevel) [reply civihr_employee_portal/src/View/ViewsResultDiff.php]: @mickadoo not currently. I tried using arguments for this but wasn't enough to get the results needed, probably because wrong integration of CiviCRM and views. So I have to use exposed filters, and as far as I know, this is the only way of achieving it, and for sure the simplest.

File: civihr_employee_portal/views/views_export/notifications/mydetails_notifications__emergency_contacts.inc (hunk 17 above)
Comment (powdevel) [reply civihr_employee_portal/views/views_export/notifications/mydetails_notifications__emergency_contacts.inc]: @mickadoo new views serves for obtaining results before and after, with that information differences are calculated. Those differences are sent in the notification email. Hence new views are created with 2 purposes: for generating differences and generating the notifications.

File: civihr_employee_portal/templates/smarty/notifications/html.tpl (hunk 16 above)
Comment (powdevel) [reply civihr_employee_portal/templates/smarty/notifications/html.tpl]: @mickadoo from another smarty template that is used to send notifications.

File: civihr_employee_portal/src/View/ViewsResultDiff.php (hunk 15 above)
Comment (powdevel) [reply civihr_employee_portal/src/View/ViewsResultDiff.php]: `ViewsDiffResultDiff` is a class that generates differences between views. I will create a class to fetch the data, and inject the results in `ViewsDiffResultDiff` to avoid external dependencies.

File: civihr_employee_portal/src/View/ViewsResultDiff.php (hunk 12 above)
Comment (powdevel) [reply civihr_employee_portal/src/View/ViewsResultDiff.php]: @mickadoo Sure. It is doable.

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php (hunk 6 above)
Comment (powdevel) [reply civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: @mickadoo I will search for code that returns this information and will use it in case I found it.

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php (hunk 3 above)
Comment (powdevel) [reply civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: @mickadoo this class is static because it does not need to store different properties in multiple instances (don't access instance data). The code for this class was approved here https://github.com/compucorp/civihr-employee-portal/pull/526 but I have to include it because it was causing conflicts with the base branch.

File: civihr_employee_portal/src/View/CustomFieldsMappingFixer.php [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+   * In this context that number is always separated by an underscore
+   *
+   * @param  string $string
+   *     the string which could have underscores or not
+   *
+   * @return string
+   *     the string without the number at the end,
+   *     if no number found at the end returns the same string
+   */
+  private static function removeLastNumberOfString($string) {
```
[the remark this answers is not in the document]
Comment (powdevel) [reply civihr_employee_portal/src/View/CustomFieldsMappingFixer.php]: @mickadoo then the name should be "removeLastNumberOfStringIfItIsAfterAnUnderscore" ?

**Comment by mickadoo** [comment]
Closing for now as unclear if this will be used
